code/opt_heuristic.py

```diff
@@ -43,48 +158,272 @@ def __init__(self, num_tries):
+        top_list = top_edge['heliostats'].to_list()
+        bottom_list = bottom_edge['heliostats'].to_list()
+        middle_list = middle['heliostats'].to_list()
+        edge_list = top_list + bottom_list
+        if len(aim_list) >= 4:
+            mid_up = df.loc[df['aimpoints']==float(1+math.ceil(len(aim_list) / 2))]
+            mid_up_list = mid_up['heliostats'].to_list()
+            middle_list += mid_up_list
+        
+        #flux,obj_val = self.switchEdgeMiddle(model,flux,helio_list,aimpoint_select,obj_val,edge_list,middle_list)
+        '''
+        mid_list,edge_list = self.createDataFrame(model,helio_list,aim_list,aimpoint_select)
+        flux,obj_val = self.switchEdgeMiddle(model,flux,helio_list,aimpoint_select,obj_val,edge_list,mid_list)
+        
+        # note: don't record or return flux as of now (or update flux model) but definitely could if wanted to
+        return(obj_val,t_end-t_start,len(defocused_helio))
     
 class SpreadHeuristic(Heuristic):
     def __init__(self):
         Heuristic.__init__(self,"spread")
     
     def getIFS(self, model):
+        t_start = time.time()
         helio_list = list(model.heliostats)
         aim_list = list(model.aimpoints)
         aimpoint_select = numpy.zeros(len(helio_list), dtype=int)
         flux = numpy.zeros(len(model.measurement_points), dtype=float)
         aim_idx = max(0, (len(aim_list) - len(helio_list)) // 2)
+        defocused_helio = []
+
         for h in range(len(helio_list)):
-            print("attempting heliostat ",h)
-            flux, a, success = self.attemptToAim(model, flux, aim_list[aim_idx], helio_list[h])
+            flux, obj_val, aim_idx, success = self.attemptToAim(model, flux, aim_list[aim_idx], helio_list[h],obj_val)
             for i in range(len(aim_list)):
                 if success: break
                 aim_idx += 1
                 aim_idx = aim_idx % len(aim_list)
                 print("attempting aimpoint ",aim_idx)    
-                flux, a, success = self.attemptToAim(model, flux, aim_list[aim_idx], helio_list[h])
+                flux, obj_val, aim_idx, success = self.attemptToAim(model, flux, aim_list[aim_idx], helio_list[h],obj_val)
             if success:
-                aimpoint_select[h] = a
+                aimpoint_select[h] = aim_idx
                 aim_idx += 1
                 aim_idx = aim_idx % len(aim_list)
-                print("heliostat #",h," aimed at ",a)
+                print("heliostat #",h," aimed at ",aim_idx)
             else: 
                 ## 
                 print("heuristic terminated at heliostat #",h)
+                defocused_helio.append(h)
                 break
-        self.updateVals(model, aimpoint_select, flux)
+        self.updateVals(model, aimpoint_select)
+        print('obj val: ',obj_val)
+        t_end = time.time()
+        return(obj_val,t_end-t_start,len(defocused_helio))
+
+
+class FluxStoreSize(Heuristic):
```

Comment:
same here - can you add docstrings in the standard format (use triple quotes, describe the function, then a list of paramters, then a list of returns)?